source/common/stats/heap_stat_data.cc

```diff
@@ -32,43 +35,24 @@ void HeapStatData::free(SymbolTable& symbol_table) {
   delete this;
 }
 
-HeapStatData& HeapStatDataAllocator::alloc(StatName name) {
-  using HeapStatDataFreeFn = std::function<void(HeapStatData * d)>;
-  std::unique_ptr<HeapStatData, HeapStatDataFreeFn> data_ptr(
-      HeapStatData::alloc(name, symbolTable()),
-      [this](HeapStatData* d) { d->free(symbolTable()); });
-  Thread::ReleasableLockGuard lock(mutex_);
-  auto ret = stats_.insert(data_ptr.get());
-  HeapStatData* existing_data = *ret.first;
-  lock.release();
-
-  if (ret.second) {
-    return *data_ptr.release();
-  }
-  ++existing_data->ref_count_;
-  return *existing_data;
-}
-
 void HeapStatDataAllocator::free(HeapStatData& data) {
-  ASSERT(data.ref_count_ > 0);
-  if (--data.ref_count_ > 0) {
-    return;
-  }
-
+  ASSERT(data.ref_count_ == 0);
   {
     Thread::LockGuard lock(mutex_);
-    size_t key_removed = stats_.erase(&data);
-    ASSERT(key_removed == 1);
+    size_t count = counters_.erase(data.statName()) + gauges_.erase(data.statName());
```

Comment:
nit: const

Also, it seems strange to me that we are erasing from both counters and gauges and expect only a single erase. What if somehow we have a counter and gauge with the same name. Can you clarify the semantics here?

Reply:
I should probably just move the cleanup into the CounterImpl and GaugeImpl destructors so I don't have to guess which map to erase from.